File: 2685. Count the Number of Complete Components/Solution.py
class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        uf = UFDS(n)
        p = set()
        for x, y in edges: uf.union(x, y)
        for i in range(n):
            q = uf.find(i)
            if q in p: continue
            if uf.isCCC(q): p.add(q)
        return len(p)

class UFDS:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.rank = [0 for i in range(n)]
        self.size = [1 for i in range(n)]
        self.edges = [0 for i in range(n)]
    
    def find(self, x):
        if self.parent[x] != x: self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def same(self, x, y):
        return self.find(x) == self.find(y)

    def union(self, x, y):
        if self.same(x, y):
            self.edges[self.find(x)] += 1
            return
        a, b = self.find(x), self.find(y)
        if self.rank[a] > self.rank[b]:
            self.parent[b] = a
            self.size[a] += self.size[b]
            self.edges[a] += self.edges[b] + 1
        else:
            self.parent[a] = b
            self.size[b] += self.size[a]
            self.edges[b] += self.edges[a] + 1
            if self.rank[a] == self.rank[b]: self.rank[b] += 1
    
    def isCCC(self, x):
        return self.edges[x] == self.size[x] * (self.size[x] - 1) // 2

File: 2685. Count the Number of Complete Components/Solution.java
class Solution {
    public int countCompleteComponents(int n, int[][] edges) {
        UFDS uf = new UFDS(n);
        Set<Integer> p = new HashSet<>();
        for (int[] e : edges) uf.union(e[0], e[1]);
        for (int i = 0; i < n; i++) {
            int q = uf.find(i);
            if (p.contains(q)) continue;
            if (uf.isCCC(q)) p.add(q);
        }
        return p.size();
    }
}

class UFDS {
    int[] parent, rank, size, edges;
    
    public UFDS(int n) {
        parent = new int[n];
        rank = new int[n];
        size = new int[n];
        edges = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }
    
    public int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }
    
    public boolean same(int x, int y) {
        return find(x) == find(y);
    }
    
    public void union(int x, int y) {
        if (same(x, y)) {
            edges[find(x)]++;
            return;
        }
        int a = find(x), b = find(y);
        if (rank[a] > rank[b]) {
            parent[b] = a;
            size[a] += size[b];
            edges[a] += edges[b] + 1;
        } else {
            parent[a] = b;
            size[b] += size[a];
            edges[b] += edges[a] + 1;
            if (rank[a] == rank[b]) rank[b]++;
        }
    }
    
    public boolean isCCC(int x) {
        return edges[x] == size[x] * (size[x] - 1) / 2;
    }
}